src/RwaRegistry.sol

```diff
@@ -0,0 +1,563 @@
+     * @param ilk The ilk name.
+     */
+    event AddDeal(bytes32 indexed ilk);
+
+    /**
+     * @notice The deal identified by `ilk` was finalized.
+     * @param ilk The ilk name.
+     */
+    event FinalizeDeal(bytes32 indexed ilk);
+
+    /**
+     * @notice Supported component `component` was added to the registry.
+     * @param component The new supported component name.
+     */
+    event AddSupportedComponent(bytes32 indexed component);
+
+    /**
+     * @notice A Deal component for `ilk` was set.
+     * @param ilk The ilk name.
+     * @param name The component name.
+     * @param addr The component address.
+     * @param variant The component variant.
+     */
+    event SetComponent(bytes32 indexed ilk, bytes32 indexed name, address addr, uint8 variant);
+
+    /**
+     * @notice Revert reason when `msg.sender` does not have the required admin access.
+     */
+    error Unauthorized();
+
+    /**
+     * @notice Revert reason when trying to add an ilk which already exists.
+     * @param ilk The ilk related to the deal being added.
+     */
+    error DealAlreadyExists(bytes32 ilk);
+
+    /**
+     * @notice Revert reason when trying to read or modify a deal for an ilk which does not exist.
+     * @param ilk The ilk related to the deal being added.
+     */
+    error DealDoesNotExist(bytes32 ilk);
+
+    /**
+     * @notice Revert reason when trying to modify a deal which was already finalized.
+     * @param ilk The ilk related to the deal being added.
+     */
+    error DealIsNotActive(bytes32 ilk);
+
+    /**
+     * @notice Revert reason when trying to add a component with address set to `address(0)`.
+     * @param ilk The ilk related to the deal being added.
+     * @param name The component name.
+     */
+    error InvalidComponentAddress(bytes32 ilk, bytes32 name);
+
+    /**
+     * @notice Revert reason when trying to add an unsupported component.
+     * @param name The unsupported component name.
+     */
+    error UnsupportedComponent(bytes32 name);
+
+    /**
+     * @notice Revert reason when trying to add an supported component more than once.
+     * @param name The component name.
+     */
+    error ComponentAlreadySupported(bytes32 name);
+
+    /**
+     * @notice Revert reason when trying to add components with mismatching params.
+     */
+    error MismatchingComponentParams();
+
+    /**
+     * @notice Revert reason when calling iterator methods with bad parameters.
+     */
+    error InvalidIteration();
+
+    /**
+     * @notice Revert reason when trying to get a component `name` which does not exist for the deal identified by `ilk`
+     * @param ilk The ilk name.
+     * @param name The unsupported component name.
+     */
+    error ComponentDoesNotExist(bytes32 ilk, bytes32 name);
+
+    /**
+     * @notice Only addresses with admin access can call methods with this modifier.
+     */
+    modifier auth() {
+        if (wards[msg.sender] != 1) {
+            revert Unauthorized();
```

Comment:
Maker code should be generally patterned after `dss` core code. The wards pattern is one of the most fundamental patterns in the Maker codebase and we should avoid attempts to innovate here as it creates new interface patterns that need to be worked around.

Example from the Clipper.sol

```
// --- Auth ---
    mapping (address => uint256) public wards;
    function rely(address usr) external auth { wards[usr] = 1; emit Rely(usr); }
    function deny(address usr) external auth { wards[usr] = 0; emit Deny(usr); }
    modifier auth {
        require(wards[msg.sender] == 1, "Clipper/not-authorized");
        _;
    }
```